provider/datagen/src/registry.rs

```diff
@@ -404,6 +404,66 @@ registry!(
     icu_timezone::provider::MetazonePeriodV1Marker = "time_zone/metazone_period@1",
 );
 
+pub use heap_measurements::MeasuringAllocator;
+
+// Inspired by the assert_no_alloc crate
+mod heap_measurements {
+    use std::alloc::{GlobalAlloc, Layout, System};
+    use std::cell::Cell;
+
+    thread_local! {
+        static ACTIVE: Cell<bool> = Cell::new(false);
+        static HEAP_SIZE_CHANGE: Cell<i64> = Cell::new(0);
+        static MAX_HEAP_SIZE: Cell<u64> = Cell::new(0);
+    }
+
+    pub fn measure<T, F: FnOnce() -> T>(func: F) -> (T, (i64, u64)) {
+        // RAII guard for managing the alloc zone.
+        struct Guard;
+        impl Guard {
+            fn new() -> Self {
+                ACTIVE.with(|c| c.set(true));
+                HEAP_SIZE_CHANGE.with(|c| c.set(0));
+                MAX_HEAP_SIZE.with(|c| c.set(0));
+                Self
+            }
+
+            fn count(self) -> (i64, u64) {
+                ACTIVE.with(|c| c.set(false));
+                (HEAP_SIZE_CHANGE.with(|c| c.get()), MAX_HEAP_SIZE.with(|c| c.get()))
+            }
+        }
+        impl Drop for Guard {
+            fn drop(&mut self) {
+                ACTIVE.with(|c| c.set(false));
+            }
+        }
+        let guard = Guard::new();
+        (func(), guard.count())
+    }
+
+    #[doc(hidden)]
+    #[derive(Debug)]
+    pub struct MeasuringAllocator;
+
+    unsafe impl GlobalAlloc for MeasuringAllocator {
+        unsafe fn alloc(&self, layout: Layout) -> *mut u8 {
+            if ACTIVE.with(|f| f.get()) {
+                HEAP_SIZE_CHANGE.with(|c| c.set(c.get() + layout.size() as i64));
+                MAX_HEAP_SIZE.with(|c| c.set(core::cmp::max(c.get() as i64, HEAP_SIZE_CHANGE.with(|c| c.get())) as u64));
```

Comment:
suggestion: we probably don't even want to track the max heap size, just the total amount of actual allocation performed; and we can track that by simply bumping this value just like we bump HEAP_SIZE_CHANGE

(call it TOTAL_BYTES_ALLOCATED)?